src/address_space.rs

```diff
@@ -99,19 +100,33 @@ impl AddressSpace {
         start: VirtualAddress,
         flags: FlagBuilder
     ) -> Result<(), &str> {
-        todo!()
+        // do not need go through the work of finding available space
+        // just test whether start address given is valid for datasource to be mapped to
+        if start + span + 2 * PAGE_SIZE < VADDR_MAX {
+            let mapping_addr = start + PAGE_SIZE;
+            let new_mapping = MapEntry::new(source, offset, span, mapping_addr, flags);
+            self.mappings.push(new_mapping);
+            self.mappings.sort_by(|a, b| a.addr.cmp(&b.addr));
+            return Ok(());
+        }
+        Err("out of address space!")
     }
 
     /// Remove the mapping to `DataSource` that starts at the given address.
     ///
     /// # Errors
     /// If the mapping could not be removed.
     pub fn remove_mapping<D: DataSource>(
-        &self,
+        &mut self,
         source: Arc<D>,
         start: VirtualAddress,
     ) -> Result<(), &str> {
-        todo!()
+        if start < VADDR_MAX {
+            let index = self.mappings.iter().position(|r| r.addr == start).unwrap();
```

Comment:
Should you `unwrap` here? If there is no such mapping, should you panic, or return an error?

A more idiomatic approach is `.position( ... ).map_err(|_| "invalid mapping")?`. The `map_err` says "if this is an `Err`, replace it with this other error, and the `?` says "if this is an `Err`, return it immediately.